### PR TITLE
Fix mouse coordinates when viewport is scrolled

### LINKED ISSUE #10190: Win32 mouse coordinates wrong when viewport scrolled
### Windows Terminal version (or Windows build number)

1.9.1445.0

### Other Software

Using Yori's mouseover support, but anything else should work.  It's just unnatural to observe this with a TUI application like FAR manager (why scroll the viewport for a full screen TUI application?)  However, scrolling the viewport should show the same behavior for anything.

### Steps to reproduce

1. Fill the console with output until it scrolls.
2. Manually scroll the viewport up.
3. Generate mouse events and observe the coordinates that the event contains.

### Expected Behavior

Since I know Dustin doesn't like the idea of applications being able to address regions that have scrolled off the visible portion of the display at the bottom, I was expecting the application to observe coordinates corresponding to mouse location if that location is visible at the bottom of the viewport.  If the cell is not visible at the bottom of the viewport, I was expecting the reported location to be capped.

### Actual Behavior

The reported mouse coordinates assume that the window has not scrolled and report coordinates on the assumption that the visible viewport is the same as the bottom of the scroll region.  This causes the application to highlight/select the wrong text, several lines below where the mouse is positioned.

terminal_mouseover

### PR DESCRIPTION
**Pull request by PankajBhojwani**
## Summary of the Pull Request
Adjust the y-coordinate of the mouse coordinates we send based on how much the viewport has been scrolled

## Validation Steps Performed
Validated: cannot repro the issue in #10190 

Closes #10190

### CODE REVIEW

File: src/cascadia/TerminalControl/ControlInteractivity.cpp [hunk 1]
```diff
@@ -182,7 +182,12 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
         else if (_canSendVTMouseInput(modifiers))
         {
-            _core->SendMouseEvent(terminalPosition, pointerUpdateKind, modifiers, 0, buttonState);
+            auto adjustment = 0;
+            if (_core->ScrollOffset() > 0)
+            {
+                adjustment = _core->BufferHeight() - _core->ScrollOffset() - _core->ViewHeight();
+            }
+            _core->SendMouseEvent({ terminalPosition.x(), terminalPosition.y() - adjustment }, pointerUpdateKind, modifiers, 0, buttonState);
```
Comment (PankajBhojwani) [inline src/cascadia/TerminalControl/ControlInteractivity.cpp:190]: Primary concern:

Can someone think of how this may unintentionally affect/break other things?

Comment (DHowett) [reply src/cascadia/TerminalControl/ControlInteractivity.cpp:190]: So, this is interesting. Inside SendMouseEvent, it locks the Y coordinate to 0. Perhaps this is correct...

It would be interesting to see what other terminal emulators do in raw mouse mode when they are scrolled up.

Comment (DHowett) [reply src/cascadia/TerminalControl/ControlInteractivity.cpp:190]: It looks like gnome-terminal, at least, completely ignores the event. It just blackholes it.

File: src/cascadia/TerminalControl/ControlInteractivity.cpp [hunk 2]
```diff
@@ -182,7 +182,12 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
         else if (_canSendVTMouseInput(modifiers))
         {
-            _core->SendMouseEvent(terminalPosition, pointerUpdateKind, modifiers, 0, buttonState);
+            auto adjustment = 0;
+            if (_core->ScrollOffset() > 0)
+            {
+                adjustment = _core->BufferHeight() - _core->ScrollOffset() - _core->ViewHeight();
+            }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlInteractivity.cpp:189]:
```suggestion
            const auto adjustment = _core->ScrollOffset() > 0 ? _core->BufferHeight() - _core->ScrollOffset() - _core->ViewHeight() : 
                                                                0;
```
If you use a ternary operator, you can throw a `const` in there

File: src/cascadia/TerminalControl/ControlInteractivity.cpp (hunk 1 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/ControlInteractivity.cpp:190]: If you want, you could future-proof this for [Infinite Scrollback](https://github.com/microsoft/terminal/issues/1410) by `std::max(terminal.Position.y() - adjustment, 0)`? That's the only thing I can think of.

**Review by carlos-zamora — APPROVED** [review]

**Review by DHowett — CHANGES_REQUESTED** [review]
blocking for discussion about what happens to out of bounds coords

File: src/cascadia/TerminalControl/ControlInteractivity.cpp (hunk 2 above)
Comment (lhecker) [reply src/cascadia/TerminalControl/ControlInteractivity.cpp:189]: Alternatively you can drop the `const` on `terminalPosition` and just modify the Y value in place within the if condition.

File: src/cascadia/TerminalControl/ControlInteractivity.cpp (hunk 1 above)
Comment (PankajBhojwani) [reply src/cascadia/TerminalControl/ControlInteractivity.cpp:190]: We follow gnome terminal! Clicks outside the active region are ignored

File: src/cascadia/TerminalControl/ControlInteractivity.cpp [hunk 3]
```diff
@@ -182,12 +182,12 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
         else if (_canSendVTMouseInput(modifiers))
         {
-            auto adjustment = 0;
-            if (_core->ScrollOffset() > 0)
+            const auto adjustment = _core->ScrollOffset() > 0 ? _core->BufferHeight() - _core->ScrollOffset() - _core->ViewHeight() : 0;
+            // If the click happened outside the active region, just don't send any mouse event
+            if (const auto adjustedY = terminalPosition.y() - adjustment >= 0)
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlInteractivity.cpp:187]: Wow, this is hard to parse. Can you separate it up like this?

```suggestion
            if (const auto adjustedY = terminalPosition.y() - adjustment; adjustedY >= 0)
```

Comment (PankajBhojwani) [reply src/cascadia/TerminalControl/ControlInteractivity.cpp:187]: Fixed!

**Comment by miniksa** [comment]
> blocking for discussion about what happens to out of bounds coords

I'm fine with out of bounds coordinates blackholing since it was said that it matches the known behavior of other terminals.

**Comment by PankajBhojwani** [comment]
> I'm fine with out of bounds coordinates blackholing since it was said that it matches the known behavior of other terminals.

Yep, thats what this PR does

**Review by DHowett — CHANGES_REQUESTED** [review]
Hate to be a pill... but since we can now write tests for ControlInteractivity, is there a way you can write a test for this? :smile:

File: src/cascadia/TerminalControl/ControlInteractivity.cpp [hunk 4]
```diff
@@ -182,7 +182,12 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
         else if (_canSendVTMouseInput(modifiers))
         {
-            _core->SendMouseEvent(terminalPosition, pointerUpdateKind, modifiers, 0, buttonState);
+            const auto adjustment = _core->ScrollOffset() > 0 ? _core->BufferHeight() - _core->ScrollOffset() - _core->ViewHeight() : 0;
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlInteractivity.cpp:185]: I'm surprised we don't already have a thing that calculates "how far back from the top line are we scrolled"... hmm.

**Comment by PankajBhojwani** [comment]
> Hate to be a pill... but since we can now write tests for ControlInteractivity, is there a way you can write a test for this? 😄

Test added!

File: src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp [hunk 5]
```diff
@@ -544,4 +545,93 @@ namespace ControlUnitTests
...
+        // move the mouse as if to make a selection
+        const til::point terminalPosition1{ 10, 4 };
+        const til::point cursorPosition1 = terminalPosition1 * fontSize;
+        interactivity->PointerMoved(leftMouseDown,
+                                    WM_LBUTTONDOWN, //pointerUpdateKind
+                                    modifiers,
+                                    true, // focused,
+                                    cursorPosition1);
+        Log::Comment(L"Verify that there's still no selection");
+        VERIFY_IS_FALSE(core->HasSelection());
```
Comment (DHowett) [inline src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp:635]: we may not have a way to test that it actually doesn't generate mouse events. oh well :smile:

@zadjii-msft any ideas? would we need the mock connection to tell us? that sounds tedious.

**Review by DHowett — APPROVED** [review]

File: src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp (hunk 5 above)
Comment (zadjii-msft) [reply src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp:635]: ew yea that would probably require a lot more plumbing than it's worth.
